src/execution/mod.rs

```diff
@@ -342,6 +344,260 @@ where
+                            let mut stack = Stack::new();
+                            // TODO: fully implement run_const
+                            run_const(wasm, &mut stack, (), &[]);
+                            let value = stack.peek_unknown_value();
+                            if value.is_none() {
+                                panic!("No value on the stack for element segment offset");
+                            }
+                            value.unwrap()
+                        };
+
+                        // TODO: this shouldn't be a simple value, should it? I mean it can't be, but it can also be any type of ValType
+                        // TODO: also, do we need to forcefully make it i32?
+                        let offset: u32 = match value {
+                            Value::I32(val) => val,
+                            Value::I64(val) => {
+                                if val > u32::MAX as u64 {
+                                    panic!("i64 value for data segment offset is out of reach")
+                                }
+                                val as u32
+                            }
+                            // INFO: no need to implement all of them, it's either i32 or i64, otherwise offset is WRONG
+                            // INFO2: wait, we might need globals handling, now that I think about it, but make it return an u32 anyways
+                            _ => unreachable!(),
+                        };
+                        let offset: usize = offset as usize;
+
+                        let el = match el.ty() {
+                            crate::RefType::FuncRef => {
+                                ElemInst {
+                                    ty: el.ty(),
+                                    elem: match &el.init {
+                                        // ElemItems::Exprs(_, _) => unreachable!(),
+                                        ElemItems::Exprs(_, exprs) => {
+                                            (*exprs).iter().map(|expr| {
+                                                let value = {
+                                                    let mut wasm = WasmReader::new(validation_info.wasm);
+                                                    wasm.move_start_to(*expr).unwrap_validated();
+                                                    let mut stack = Stack::new();
+                                                    // TODO: fully implement run_const
+                                                    run_const(wasm, &mut stack, (), &[]);
+                                                    let value = stack.peek_unknown_value();
+                                                    if value.is_none() {
+                                                        panic!("No value on the stack for element segment offset");
+                                                    }
+                                                    value.unwrap()
```

Comment:
I'm not going to write out all "Ditto panic"s here, you get the idea. Also, I failed to notice until now the `.unwrap()`. May I suggest `.unwrap_validated()`